### PR TITLE
fix(Core/Spells): Implemented spell priorities.

### LINKED ISSUE #11261: Mage T2 8/8 set bonus wrongly applied
### What client do you play on?

enUS

### Faction

- [x] Alliance
- [X] Horde

### Content Phase:

- [ ] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59
- [X] 60

### Current Behaviour

Set bonus stacks wrong with some talents:
1. Brain Freeze - endless cast of fireball when u get both procs
2. Hot Streak - consumes both procs at the same time 

### Expected Blizzlike Behaviour

1. Cast time shouldn't be negative. Only one proc must be spent for each cast
2. Only one proc must be spent for each cast

### Steps to reproduce the problem

1. Get both procs at the same time and try to cast fireball(for Brain Freeze) or Pyroblast(for Hot Streak )

### Extra Notes
Originally reported: https://github.com/chromiecraft/chromiecraft/issues/3268

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/856aed6fc6ad34fd07f1aa9df2a39c710ffd5b33

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-bg-item-reward](https://github.com/azerothcore/mod-bg-item-reward)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chat-transmitter](https://github.com/azerothcore/mod-chat-transmitter)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-cta-switch](https://github.com/azerothcore/mod-cta-switch)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-low-level-rbg](https://github.com/azerothcore/mod-low-level-rbg)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-rdf-expansion](https://github.com/azerothcore/mod-rdf-expansion)
- [mod-progression-system](https://github.com/azerothcore/mod-progression-system)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-exchange-npc](https://github.com/55Honey/Acore_ExchangeNpc)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)
- [lua-zone-debuff](https://github.com/55Honey/Acore_ZoneDebuff)

### Customizations

None

### Server

ChromieCraft

### LINKED ISSUE #11273: Warriors T2 Battlegear of Wrath 5/8 not working correctly with Talent
### What client do you play on?

enUS

### Faction

- [x] Alliance
- [X] Horde

### Content Phase:

- [ ] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59
- [X] 60

### Current Behaviour

5 pieces: 20% chance after using an offensive ability requiring rage that your next offensive ability requires 5 less rage to use.
Wrong stacking with [Sword and Board](https://wowgaming.altervista.org/aowow/?spell=50227) talent proc. When the talent and set bonuse proc, your next shield slam ability eats both of them.

### Expected Blizzlike Behaviour

Set bonuse [proc](https://wowgaming.altervista.org/aowow/?spell=21890) should consumes only by an ability that requires rage. So proc and talent should not be spent at the same time, but in some sequence.

### Source

I don't have any proof and I'm pretty sure it's impossible to find at the moment. I rely on similar mechanics of talents and various procs on other classes. So maybe I'm wrong. But it looks more logical to be honest. 

### Steps to reproduce the problem

1. Get a talent and set bonus procs at the same time 
2. Use your Shield Slam ability

### Extra Notes
Originally reported: https://github.com/chromiecraft/chromiecraft/issues/3227

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/e8d3292533c7e5d18af15d9bffabd581bacbadf9

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-bg-item-reward](https://github.com/azerothcore/mod-bg-item-reward)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chat-transmitter](https://github.com/azerothcore/mod-chat-transmitter)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-cta-switch](https://github.com/azerothcore/mod-cta-switch)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-low-level-rbg](https://github.com/azerothcore/mod-low-level-rbg)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-rdf-expansion](https://github.com/azerothcore/mod-rdf-expansion)
- [mod-progression-system](https://github.com/azerothcore/mod-progression-system)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-exchange-npc](https://github.com/55Honey/Acore_ExchangeNpc)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)
- [lua-zone-debuff](https://github.com/55Honey/Acore_ZoneDebuff)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by UltraNix**
Fixes #11261

## Issues Addressed:

- Closes #11261
- Closes #11273
- Closes https://github.com/chromiecraft/chromiecraft/issues/3268

## Tests Performed:

- Not tested.

## How to Test the Changes:

Get both procs at the same time and try to cast fireball(for Brain Freeze) or Pyroblast(for Hot Streak )
Only one of the procs is consumed per cast.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Nefertumm** [comment]
build issues

**Comment by UltraNix** [comment]
Done.

**Comment by fputs** [comment]
Tested with Pryoblast. In all of the following tests, Netherwind Focus was consumed before Hot Streak:

Both procs, cast Fireball to consume Netherwind Focus, cast Pyroblast to consume Hot Streak
Both procs, cast Pyroblast to consume Netherwind Focus, cast Pyroblast to consume Hot Streak
Both procs, cast Fireball to consume Netherwind Focus, cast Fireball again, cast Pyroblast to consume Hot Streak

**Comment by fputs** [comment]
Tested with Brain Freeze and failed. When both buffs are on the Mage, casting an instant Fireball consumes both Neterwind Focus and Brain Freeze. I was able to reproduce this behavior 3 times in a row

**Comment by UltraNix** [comment]
@fputs Test it again.

**Comment by fputs** [comment]
Tested new commits.

[x] Netherwind Focus was consumed before Hot Streak on consecutive Pyroblasts.
[x] Brain Freeze was consumed before Netherwind Focus on consecutive Fireballs.

File: src/server/game/Spells/SpellInfo.cpp [hunk 1]
```diff
@@ -831,6 +831,7 @@ SpellInfo::SpellInfo(SpellEntry const* spellEntry)
     SpellVisual = spellEntry->SpellVisual;
     SpellIconID = spellEntry->SpellIconID;
     ActiveIconID = spellEntry->ActiveIconID;
+    Priority = spellEntry->SpellPriority;
```
Comment (Kitzunu) [inline src/server/game/Spells/SpellInfo.cpp:834]:
```suggestion
    SpellPriority = spellEntry->SpellPriority;
```

File: src/server/game/Spells/SpellInfo.h [hunk 2]
```diff
@@ -377,6 +377,7 @@ friend class SpellMgr;
     std::array<uint32, 2> SpellVisual;
     uint32 SpellIconID;
     uint32 ActiveIconID;
+    uint32 Priority;
```
Comment (Kitzunu) [inline src/server/game/Spells/SpellInfo.h:380]:
```suggestion
    uint32 SpellPriority;
```

File: src/server/game/Entities/Player/Player.h [hunk 3]
```diff
@@ -2920,65 +2922,117 @@ template <class T> T Player::ApplySpellMod(uint32 spellId, SpellModOp op, T& bas
...
+        spell = m_spellModTakingSpell;
     }
+
+    SpellModifier* chargedMod = nullptr;
+    for (auto mod : m_spellMods[op])
+    {
+        // Charges can be set only for mods with auras
+        if (!mod->ownerAura)
+        {
+            ASSERT(mod->charges == 0);
```
Comment (Kitzunu) [inline src/server/game/Entities/Player/Player.h:3000]: Why not ASSERT(!mod->charges);

**Comment by UltraNix** [comment]
@Kitzunu Done.

**Review by Kitzunu — APPROVED** [review]

**Review by eziohamed — APPROVED** [review]

**Pull request merged**